Add JOINs piece to relational databases section (#470)
prisma/dataguide: content/03-types/02-relational/07-what-are-joins.mdx

Let’s demonstrate this concept with a table holding customer information. In the table below, the columns, or customer attributes, are `name`, `age`, and `email address` and the rows represent each customer’s `name`, `age`, and `email address`.

| name    | age | email address     |
| ------- | --- | ----------------- |
| Jenny   | 30  | jenny@prisma.io   |
| Adrian  | 28  | adrian@prisma.io  |
| Synniva | 25  | synniva@prisma.io |

A relational database has the ability to define relationships, or connections, between different tables using primary and foreign keys. All tables in a relational database have a column known as the _primary key_ which uniquely identifies each row. To see how this works in practice, let’s use our previous example. The primary key for the `customers` table would be a column of unique IDs for each customer, called `customer_id`. In other words, no customer would share the same id as another.

> ![table-properties](../../dataguide-images/types/relational/what-are-joins/table-properties.png)

Author: updated image to say column and row instead of attribute and record and included the green highlight, thank you for catching that 🙏

A _foreign key_ is used to create a relationship between tables by referencing the primary key of another table. To demonstrate the concept of foreign keys, let’s say we have another table called `orders` with columns: `order_id`, `cost`, and `order_date`. We can link the `orders` table (left table in diagram) with the `customers` table (right table in diagram) by including a `customer_id` column (foreign key), associating each order with the corresponding customer.

> ![relationship-between-tables](../../dataguide-images/types/relational/what-are-joins/relationship-between-tables.png)

Author: It seems it keeps getting compressed even though I've made it bigger, hoping to get that fixed soon

The basic syntax of a _JOIN_ operation is as follows:

```sql
SELECT
    *
FROM
    <first_table>
<join_type> <second_table>
    <join_condition>
```

Going top-down, the first part of the query specifies which columns you would like to include from the resulting dataset. In our case, the `SELECT` selects all (`*`) columns from the resulting dataset. In the `FROM` clause, you specify the source table from which data will be retrieved, which is the first table. Afterwards, you specify the specific type of JOIN operation you want to perform with the second table. Lastly, the JOIN condition represents how the two tables should be joined. It typically consists of equality comparison between columns from the two tables.

Using our example from earlier, when you perform a standard JOIN between the `customers` and `orders` tables, the database looks for matching values in the `customer_id` column of the `orders` table (foreign key) and the `customer_id` column of the `customers` table (primary key). It then combines the rows where these values match into a single result set.

The result set for the tables would be:

> | customer_id | name    | age | email address     | order_id | cost    | order_date |

Author: [7184537](https://github.com/prisma/dataguide/pull/470/commits/7184537c837e4d2343302691a1ce420c45ddc79f)

| ----------- | ------- | --- | ----------------- | -------- | ------- | ---------- |
| 1           | Jenny   | 30  | jenny@prisma.io   | 1        | $150.00 | 4/1/2024   |
| 2           | Adrian  | 28  | adrian@prisma.io  | 2        | $200.00 | 3/29/2024  |
| 1           | Jenny   | 30  | jenny@prisma.io   | 3        | $20.00  | 4/2/2024   |
| 3           | Synniva | 25  | synniva@prisma.io | 4        | $320.00 | 4/4/2024   |

The result set combines the columns from both the `orders` table and `customers` table based on the matching rows from the `customer_id` column. This allows you to see order information alongside the customer information.

> ## Different Types of JOINs in MySQL and PostgreSQL

Author: The only differences from the two (at least from the data guide sources I used) are that MySQL combines Cross Join and Inner Join and Full join is implemented differently. I included a note for that here:
![image](https://github.com/prisma/dataguide/assets/74070079/7d1c7f6a-1a41-45dd-8bda-5a915a1e26ec)
For the full join I included a separate code snippet. 

I agree the naming of the heading doesn't really make sense. In that case, I'm thinking of removing "Different" from the heading. Do you think that would make more sense?

### Traditional JOINs

The most common form of JOIN operations you will encounter are `INNER JOIN`, `LEFT JOIN`, `RIGHT JOIN`, `FULL JOIN`, and `CROSS JOIN`. To demonstrate the following types of JOINs, we will use the following `customers` and `orders` (modified) tables:

| customer_id | name    | age | email address     |
| ----------- | ------- | --- | ----------------- |
| 1           | Jenny   | 30  | jenny@prisma.io   |
| 2           | Adrian  | 28  | adrian@prisma.io  |
| 3           | Synniva | 25  | synniva@prisma.io |

| order_id(primary key) | cost    | order_date | customer_id(foreign key) |
| --------------------- | ------- | ---------- | ------------------------ |
| 1                     | $150.00 | 4/1/2024   | 1                        |
| 2                     | $200.00 | 3/29/2024  | 2                        |
| 3                     | $20.00  | 4/2/2024   | 1                        |

#### **INNER JOIN**

The `INNER JOIN` is the default JOIN and returns rows from both tables only where there is a match. Here is the associated query for the `customers` and `orders` table:

```sql
SELECT
    *
FROM
    customers

> [INNER] JOIN orders

Author: it was initially showing the syntax but you're right it's best removed

    ON customers.customer_id = orders.customer_id;
```

Here is the result table after using an `INNER JOIN`:

| customer_id | name   | age | email address    | order_id | cost    | order_date |
| ----------- | ------ | --- | ---------------- | -------- | ------- | ---------- |
| 1           | Jenny  | 30  | jenny@prisma.io  | 1        | $150.00 | 4/1/2024   |
| 1           | Jenny  | 30  | jenny@prisma.io  | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian | 28  | adrian@prisma.io | 2        | $200.00 | 3/29/2024  |

When depicted as a Venn diagram, an `INNER JOIN` represents the overlapping region of the two circles. In other words, only the values existing in both tables are included.

> ![inner-join](../../dataguide-images/types/relational/what-are-joins/inner-join.png)

Reviewer: for this image and following (left, right, full join), can we maybe make them a bit smaller? They feel disproportionately big compared to the rest. Maybe a 50-70% size?

| 1           | Jenny  | 30  | jenny@prisma.io  | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian | 28  | adrian@prisma.io | 2        | $200.00 | 3/29/2024  |

In this case, the result is the same as an `INNER JOIN` as each order has a corresponding customer.

When depicted as a Venn diagram, a `RIGHT JOIN` represents the entire right circle. In other words, the matching values from both tables will be included along with all records from the `orders` table.

![right-join](../../dataguide-images/types/relational/what-are-joins/right-join.png)

#### **FULL JOIN**

A `FULL JOIN` returns all rows when there is a match in either table. Here is the associated query for the `customers` and `orders` table in PostgreSQL:

```sql
SELECT
    *
FROM
    customers
FULL JOIN orders
    ON customers.customer_id = orders.customer_id;
```

MySQL does not natively support `FULL JOIN`s. As a workaround, we can use a `LEFT JOIN` combined with an “anti-JOIN,” or a JOIN operation that finds results that are _not_ in common between tables (specified by the NULL). The `UNION ALL` allows us to combine these together.

```sql

> (   SELECT
>         *
>     FROM
>         customers
>     LEFT JOIN orders
>         ON customers.customer_id = orders.customer_id
> )
> UNION ALL
> (   SELECT
>         *
>     FROM
>         customers
>     RIGHT JOIN orders
>         ON customers.customer_id = orders.customer_id
>     WHERE customers.customer_id IS NULL
> );

Author: ```suggestion
  (SELECT *
   FROM customers
   LEFT JOIN orders ON customers.customer_id = orders.customer_id)
UNION ALL
  (SELECT *
   FROM customers
   RIGHT JOIN orders ON customers.customer_id = orders.customer_id
   WHERE customers.customer_id IS NULL );
```

```

Here is the result table after using a `FULL JOIN`:

| customer_id | name    | age | email address     | order_id | cost    | order_date |
| ----------- | ------- | --- | ----------------- | -------- | ------- | ---------- |
| 1           | Jenny   | 30  | jenny@prisma.io   | 1        | $150.00 | 4/1/2024   |
| 1           | Jenny   | 30  | jenny@prisma.io   | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian  | 28  | adrian@prisma.io  | 2        | $200.00 | 3/29/2024  |
| 3           | Synniva | 25  | synniva@prisma.io | NULL     | NULL    | NULL       |

In this case, the result would be the same as a `LEFT JOIN` as every record from the `customers` table has a match in the `orders` table.

When depicted as a Venn diagram, a `FULL JOIN` represents both circles. In other words, a `FULL JOIN` combines all records from both tables.

![full-join](../../dataguide-images/types/relational/what-are-joins/full-join.png)

#### **CROSS JOIN**

> A `CROSS JOIN` returns the Cartesian product of the two tables, meaning every row from the first table is combined with every row from the second table. In this syntax, the result is formed by adding each of the rows in the first table with each of the rows from the second table like so:

Author: ```suggestion
A `CROSS JOIN` returns the [Cartesian product](https://www.sciencedirect.com/topics/computer-science/cartesian-product) of the two tables, meaning every row from the first table is combined with every row from the second table. In this syntax, the result is formed by adding each of the rows in the first table with each of the rows from the second table like so:
```

added link to sciencedirect as I found the explanation to be clearer

```sql
t1.r1 + t2.r1
t1.r1 + t2.r2
t1.r1 + t2.r3
t1.r2 + t2.r1
t1.r2 + t2.r2
t1.r2 + t2.r3
t1.r3 + t2.r1
t1.r3 + t2.r2
t1.r3 + t2.r3
```

> **Note:** In MySQL, the concept of a `CROSS JOIN` is combined with the `INNER JOIN`. Read more in the [data guide](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables#different-types-of-joins).

With a `CROSS JOIN`, each row from the `customers` table is combined with each row from the `orders` table, resulting in 9 rows. However, we won’t show the result table as a `CROSS JOIN` wouldn’t accurately pair the customer with their respective orders.

> To learn more about JOINs in PostgreSQL, read more in the data guide [here](https://www.prisma.io/dataguide/postgresql/reading-and-querying-data/joining-tables). To learn more about JOINs in MySQL, read more in the data guide [here](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables).

Author: definitely a much better wording thank you 🙂

### LATERAL JOINs

Lateral JOINs offer a different syntax to combining tables compared to traditional JOINs. In a `LATERAL JOIN`, the second table is presented as a subquery, and the JOIN criteria is defined within the WHERE clause of the subquery.

According to the [PostgreSQL](https://www.postgresql.org/docs/9.3/sql-select.html#SQL-FROM) docs, “The `LATERAL` key word can precede a sub-`SELECT FROM` item. This allows the sub-`SELECT` to refer to columns of `FROM` items that appear before it in the `FROM` list. (Without `LATERAL`, each sub-`SELECT` is evaluated independently and so cannot cross-reference any other `FROM` item.)”

> > **Note:** `LATERAL JOIN`s are supported by Postgres but only supported in [MySQL versions >8](https://dev.mysql.com/blog-archive/support-for-lateral-derived-tables-added-to-mysql-8-0-14/)

Reviewer 2: ```suggestion
> **Note:** `LATERAL JOIN`s are supported by Postgres but only supported in [MySQL versions >8](https://dev.mysql.com/blog-archive/support-for-lateral-derived-tables-added-to-mysql-8-0-14/).
```

Author: Ok, I wasn't able to find the suggestion to turn the postgres quote into a markdown quote but I assume you meant something like this: 

According to the [PostgreSQL](https://www.postgresql.org/docs/9.3/sql-select.html#SQL-FROM) docs:
> “The `LATERAL` key word can precede a sub-`SELECT FROM` item. This allows the sub-`SELECT` to refer to columns of `FROM` items that appear before it in the `FROM` list. (Without `LATERAL`, each sub-`SELECT` is evaluated independently and so cannot cross-reference any other `FROM` item.)”

I moved the note to the bottom of the lateral joins section (also right before the Prisma outlink but I thought that was probably better than how it is currently): [7260f3c](https://github.com/prisma/dataguide/pull/470/commits/7260f3c8940e27d24bad56a6670a2ac9600e62c3)

In other words, a `LATERAL JOIN` is like a foreach loop, where PostgreSQL iterates over each row in a result set and uses each row to evaluate the subquery.

Here is the associated query for the `customers` and `orders` table:

> ```sql
> SELECT
>     *
> FROM
>     customers
> 	LEFT JOIN LATERAL
> 	    (SELECT *
>        FROM orders
>        WHERE orders.customer_id = customers.customer_id
>        ) AS alias ON TRUE;
> 
> ```

Reviewer 2: proper formatting according to https://sqlformat.org:

```suggestion
```sql
SELECT *
FROM customers
LEFT JOIN LATERAL
  (SELECT *
   FROM orders
   WHERE orders.customer_id = customers.customer_id ) AS ALIAS ON TRUE;
```
```

Here is the result table after using a `LATERAL JOIN`:

| customer_id | name    | age | email address     | order_id | cost    | order_date |
| ----------- | ------- | --- | ----------------- | -------- | ------- | ---------- |
| 1           | Jenny   | 30  | jenny@prisma.io   | 1        | $150.00 | 4/1/2024   |
| 1           | Jenny   | 30  | jenny@prisma.io   | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian  | 28  | adrian@prisma.io  | 2        | $200.00 | 3/29/2024  |
| 3           | Synniva | 25  | synniva@prisma.io | NULL     | NULL    | NULL       |

In this case, the result would be the same as a `LEFT JOIN` and `FULL JOIN`.

<PrismaOutlinks>

> 💡 Prisma ORM’s preview JOIN feature works with Postgres and MySQL versions >8. To read more about the preview feature, read the blog post [here](https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview)

Reviewer 2: ```suggestion
💡 Prisma ORM makes it easy to query relations between tables without thinking about the intricacies and low-level complexities of SQL operations. Read the docs to learn more about [relation queries](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries).
```

</PrismaOutlinks>

## Implementing JOINs for different relations

> It’s important to understand database relations when selecting the appropriate JOIN type. In the context of databases, relations describe a relationship between tables in a database. 

Reviewer 2: ```suggestion
It’s important to understand database _relations_ when selecting the appropriate JOIN type. In the context of databases, relations describe a relationship between tables in a database. 
```

In this section, we will illustrate both the one-to-one and the one-to-many relations along with considerations for selecting the appropriate JOIN type for each.

### One-to-one (1-1)

In a one-to-one relation, each record in one table is associated with exactly one record in another table, and vice versa. Here is an example SQL schema demonstrating a one-to-one relation between `User` and `Profile` models:

```sql
CREATE TABLE "User" (
    "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    "email" TEXT NOT NULL
);

CREATE TABLE "Profile" (
    "name" TEXT NOT NULL,
    "gender" TEXT NOT NULL,
    "age" TEXT NOT NULL,
    "userEmail" TEXT NOT NULL,
    CONSTRAINT "Profile_userEmail_fkey" FOREIGN KEY ("userEmail") REFERENCES "User" ("email") ON DELETE RESTRICT ON UPDATE CASCADE
);

CREATE UNIQUE INDEX "User_email_key" ON "User"("email");

CREATE UNIQUE INDEX "Profile_userEmail_key" ON "Profile"("userEmail");

```

> This means that a user can have only one profile, and a profile can belong to only one user.

Reviewer 2: I think this snippet could be confusing because it uses a unique field as a foreign key which hasn't been explained as a possibility in this article and also isn't mentioned anywhere here as an additional explanation. A reader who is new to the topic will wonder how this is possible because all previous examples used primary keys to be mapped to foreign keys.

I suggest to either use the previous approach of using a primary key or add an additional explanation.

In one-to-one relations, JOINs are straightforward. You can use a traditional `INNER JOIN` to combine the users with their profiles. Here is an example query:

```sql
SELECT

> 		user.id, user.email, profile.name, profile.gender, profile.age

Reviewer 2: any reason why we explicitly select all columns here instead of `SELECT *` here?

Author: good point. changed 👍 [2433152](https://github.com/prisma/dataguide/pull/470/commits/243315218b185c28a481afa902647a5af1f21c12)

    "email" TEXT NOT NULL
);

CREATE TABLE "Post" (
    "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    "title" TEXT NOT NULL,
    "authorId" INTEGER NOT NULL,
    CONSTRAINT "Post_authorId_fkey" FOREIGN KEY ("authorId") REFERENCES "User" ("id") ON DELETE RESTRICT ON UPDATE CASCADE
);

CREATE UNIQUE INDEX "User_email_key" ON "User"("email");

```

This means that one user can have multiple posts, but each post is linked to only one user.

When dealing with one-to-many relations, traditional JOINs can lead to duplicated data. Consider the following query:

```sql
SELECT
		user.id, user.email, post.id AS post_id, post.title AS post_title
FROM
		user
INNER JOIN post
		ON user.id = post.authorId;

```

Here is an example result set based on the query above:

| id  | email           | post_id | post_title                 |
| --- | --------------- | ------- | -------------------------- |
| 1   | jenny@prisma.io | 1       | Introduction to Prisma ORM |
| 1   | jenny@prisma.io | 2       | Introduction to JOINs      |
| 1   | jenny@prisma.io | 3       | Introduction to SQL        |
| 2   | meera@prisma.io | 4       | Using Prisma Client        |

> As you can see in the result set, Jenny's email gets duplicated over multiple rows for each of her posts. As we incorporate more nested relations, result sets may become bloated. This can make it harder to interpret the data.

Reviewer 2: not only Jenny's `email` but also the `id`

Author: 🙏  [7ab8e61](https://github.com/prisma/dataguide/pull/470/commits/7ab8e61191519b32075674502b3253276d101d2b)

To address this duplication, we can pair `LATERAL JOIN`s with JSON aggregation. JSON aggregation helps avoid duplication by consolidating multiple related records into a single field as a JSON array, while `LATERAL JOIN`s help improve readability and can optimize performance.

Here is an example of what the result set would look like using `LATERAL JOIN`s with JSON aggregation:

| id  | email           | post_titles                                                                                                                    |
| --- | --------------- | ------------------------------------------------------------------------------------------------------------------------------ |

> | 1   | jenny@prisma.io | [{"post title": "Introduction to Prisma ORM"}, {"post title": "Introduction to JOINs"}, {"post title": "Introduction to SQL"}] |
> | 2   | meera@prisma.io | [{"post title": "Using Prisma Client"}]                                                                                        |

Reviewer 2: can we use code formatting for the JSON data?

```suggestion
| 1   | jenny@prisma.io | `[{"post title": "Introduction to Prisma ORM"}, {"post title": "Introduction to JOINs"}, {"post title": "Introduction to SQL"}]` |
| 2   | meera@prisma.io | `[{"post title": "Using Prisma Client"}]`                                                                                        |
```

| 1           | Jenny   | 30  | jenny@prisma.io   | 1        | $150.00 | 4/1/2024   |
| 1           | Jenny   | 30  | jenny@prisma.io   | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian  | 28  | adrian@prisma.io  | 2        | $200.00 | 3/29/2024  |
| 3           | Synniva | 25  | synniva@prisma.io | NULL     | NULL    | NULL       |

In this case, the result would be the same as a `LEFT JOIN` as every record from the `customers` table has a match in the `orders` table.

When depicted as a Venn diagram, a `FULL JOIN` represents both circles. In other words, a `FULL JOIN` combines all records from both tables.

![full-join](../../dataguide-images/types/relational/what-are-joins/full-join.png)

#### **CROSS JOIN**

A `CROSS JOIN` returns the Cartesian product of the two tables, meaning every row from the first table is combined with every row from the second table. In this syntax, the result is formed by adding each of the rows in the first table with each of the rows from the second table like so:

```sql
t1.r1 + t2.r1
t1.r1 + t2.r2
t1.r1 + t2.r3
t1.r2 + t2.r1
t1.r2 + t2.r2
t1.r2 + t2.r3
t1.r3 + t2.r1
t1.r3 + t2.r2
t1.r3 + t2.r3
```

> <PrismaOutlinks>
> 
> In MySQL, the concept of a `CROSS JOIN` is combined with the `INNER JOIN`. Read more in the [data guide](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables#different-types-of-joins).
> 
> </PrismaOutlinks>

Author: ```suggestion
> **Note:** In MySQL, the concept of a `CROSS JOIN` is combined with the `INNER JOIN`. Read more in the [data guide](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables#different-types-of-joins).

```

With a `CROSS JOIN`, each row from the `customers` table is combined with each row from the `orders` table, resulting in 9 rows. However, we won’t show the result table as a `CROSS JOIN` wouldn’t accurately pair the customer with their respective orders.

> <PrismaOutlinks>
> 
> 💡 To learn more about JOINs in PostgreSQL, read more in the data guide [here](https://www.prisma.io/dataguide/postgresql/reading-and-querying-data/joining-tables). To learn more about JOINs in MySQL, read more in the data guide [here](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables).
> 
> </PrismaOutlinks>

Author: ```suggestion

To learn more about JOINs in PostgreSQL, read more in the data guide [here](https://www.prisma.io/dataguide/postgresql/reading-and-querying-data/joining-tables). To learn more about JOINs in MySQL, read more in the data guide [here](https://www.prisma.io/dataguide/mysql/reading-and-querying-data/joining-tables).

```

```sql
SELECT
    *
FROM
    customers
	LEFT JOIN LATERAL
	    (SELECT *
       FROM orders
       WHERE orders.customer_id = customers.customer_id
       ) AS alias ON TRUE;

```

Here is the result table after using a `LATERAL JOIN`:

| customer_id | name    | age | email address     | order_id | cost    | order_date |
| ----------- | ------- | --- | ----------------- | -------- | ------- | ---------- |
| 1           | Jenny   | 30  | jenny@prisma.io   | 1        | $150.00 | 4/1/2024   |
| 1           | Jenny   | 30  | jenny@prisma.io   | 3        | $20.00  | 4/2/2024   |
| 2           | Adrian  | 28  | adrian@prisma.io  | 2        | $200.00 | 3/29/2024  |
| 3           | Synniva | 25  | synniva@prisma.io | NULL     | NULL    | NULL       |

In this case, the result would be the same as a `LEFT JOIN` and `FULL JOIN`.

<PrismaOutlinks>

💡 Prisma ORM’s preview JOIN feature works with Postgres and MySQL versions >8. To read more about the preview feature, read the blog post [here](https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview)

</PrismaOutlinks>

## Implementing JOINs for different relations

> It’s important to understand database relations when selecting the appropriate JOIN type. In the context of databases, relations describe a relationship between tables in a database. These are defined as relations within data models in the Prisma schema.

Author: ```suggestion
It’s important to understand database relations when selecting the appropriate JOIN type. In the context of databases, relations describe a relationship between tables in a database. 
```

In this section, we will illustrate both the one-to-one and the one-to-many relations along with considerations for selecting the appropriate JOIN type for each.

> <PrismaOutlinks>

Author: Note to self: add cross-links tomorrow morning

> <PrismaOutlinks>

Author: [09ba455](https://github.com/prisma/dataguide/pull/470/commits/09ba455ec9c9741e6f852ef1196d057674b1305f)

> <PrismaOutlinks>
> 
> You can learn more about relations in Prisma ORM [here](https://www.prisma.io/docs/orm/prisma-schema/data-model/relations).
> 
> </PrismaOutlinks>

Author: ```suggestion

```
